# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [82]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import datetime
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [83]:
df = pd.read_csv('/home/vizziwo/data/step2_output.csv', index_col=0)
df = df.reset_index(drop=True)
df.head()


Name   Region    state  summit_elev  vertical_drop  \
0       Alyeska Resort   Alaska   Alaska         3939           2500   
1  Eaglecrest Ski Area   Alaska   Alaska         2600           1540   
2     Hilltop Ski Area   Alaska   Alaska         2090            294   
3     Arizona Snowbowl  Arizona  Arizona        11500           2300   
4  Sunrise Park Resort  Arizona  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  LongestRun_mi  \
0        250      1        0.0          0          2  ...            1.0   
1       1200      0        0.0          0          0  ...            2.0   
2       1796      0        0.0          0          0  ...            1.0   
3       9200      0        0.0          1          0  ...            2.0   
4       9200      0        0.0          0          1  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 27 columns]

In [84]:
df = df.drop(columns = ['Region'])

In [85]:
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  LongestRun_mi  \
0        0.0          0          2     2  ...            1.0   
1        0.0          0          0     0  ...            2.0   
2        0.0          0          0     0  ...            1.0   
3        0.0          1          0     2  ...            2.0   
4        0.0          0          1     2  ...            1.2   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   
3              777.0           104.0             122.0       81.0   
4              800.0            80.0             115.0       49.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   
3            260.0          89.0          89.0              122.0   
4            250.0          74.0          78.0              104.0   

   NightSkiing_ac  
0           550.0  
1             0.0  
2            30.0  
3             0.0  
4            80.0  

[5 rows x 26 columns]

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 330 entries, 0 to 329
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               330 non-null    object 
 1   state              330 non-null    object 
 2   summit_elev        330 non-null    int64  
 3   vertical_drop      330 non-null    int64  
 4   base_elev          330 non-null    int64  
 5   trams              330 non-null    int64  
 6   fastEight          330 non-null    float64
 7   fastSixes          330 non-null    int64  
 8   fastQuads          330 non-null    int64  
 9   quad               330 non-null    int64  
 10  triple             330 non-null    int64  
 11  double             330 non-null    int64  
 12  surface            330 non-null    int64  
 13  total_chairs       330 non-null    int64  
 14  Runs               330 non-null    float64
 15  TerrainParks       330 non-null    float64
 16  LongestRun_mi      330 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [87]:
state=df['state']
df = pd.concat([df.drop('state',axis=1), pd.get_dummies(state)], axis=1)


In [88]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [89]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler()

scaler.fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [90]:
X_scaled, y

(array([[-0.1750243 ,  1.3572862 , -1.0037286 , ..., -0.11076976,
         -0.22573306, -0.15762208],
        [-0.53401788,  0.34294523, -0.69849743, ..., -0.11076976,
         -0.22573306, -0.15762208],
        [-0.6707518 , -0.97358483, -0.50700504, ..., -0.11076976,
         -0.22573306, -0.15762208],
        ...,
        [ 0.86227679,  0.37569999,  0.91987035, ..., -0.11076976,
         -0.22573306,  6.34428877],
        [ 1.35961292, -0.23818762,  1.74270932, ..., -0.11076976,
         -0.22573306,  6.34428877],
        [ 1.31591169, -0.12196105,  1.61483353, ..., -0.11076976,
         -0.22573306,  6.34428877]]), 0      85.00000
 1      53.00000
 2      34.00000
 3      89.00000
 4      78.00000
          ...   
 325    64.16681
 326    42.00000
 327    59.00000
 328    49.00000
 329    49.00000
 Name: AdultWeekend, Length: 330, dtype: float64)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [91]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train,X_test,y_train,y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [92]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [93]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [94]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [95]:
explained_variance_score(y_test, y_pred)

0.9348640546750188

In [96]:
mean_absolute_error(y_test, y_pred) 

5.0702603568494755

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [97]:
model.intercept_

64.10355786870056

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [98]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
total_chairs       4.940910e+12
fastQuads          1.873110e+12
surface            1.754964e+12
double             1.546539e+12
New York           1.381045e+12
triple             1.379619e+12
Michigan           1.303331e+12
Colorado           1.148310e+12
California         1.123729e+12
quad               1.118130e+12
Pennsylvania       1.072333e+12
New Hampshire      9.887748e+11
Wisconsin          9.887748e+11
Vermont            9.589003e+11
Minnesota          9.278550e+11
Utah               8.955171e+11
Idaho              8.617410e+11
Montana            8.617410e+11
Massachusetts      8.263504e+11
Washington         7.891282e+11
Oregon             7.891282e+11
Maine              7.498015e+11
New Mexico         7.498015e+11
Wyoming            7.080199e+11
North Carolina     6.150645e+11
Connecticut        5.623403e+11
Ohio               5.623403e+11
fastSixes          5.552845e+11
West Virginia      5.037457e+11
Virginia           5.037457e+11
Illinois           5.037457e+11
Nevada             5.037457e+11
trams              4.771159e+11
Alaska             4.369252e+11
Iowa               4.369252e+11
New Jersey         3.572930e+11
Missouri           3.572930e+11
Indiana            3.572930e+11
Arizona            3.572930e+11
South Dakota       3.572930e+11
Tennessee          2.530291e+11
Maryland           2.530291e+11
Rhode Island       2.530291e+11
fastEight          4.690516e+10
AdultWeekday       1.930978e+01
summit_elev        1.798563e+01
base_elev          1.741354e+01
vertical_drop      5.714986e+00
Runs               1.137939e+00
NightSkiing_ac     7.980957e-01
daysOpenLastYear   3.930664e-01
LongestRun_mi      3.881836e-01
yearsOpen          3.763428e-01
averageSnowfall    3.068237e-01
Snow Making_ac     2.522481e-01
SkiableTerrain_ac  1.594238e-01
projectedDaysOpen  1.497192e-01
TerrainParks       1.373901e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

In [101]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'Alaska', 'Arizona', 'California', 'Colorado',
       'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Maine',
       'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Missouri',
       'Montana', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Dakota', 'Tennessee', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype='object')

#### Model 2

In [102]:
df2 = df.loc[:,'summit_elev':'NightSkiing_ac'].drop(columns=['AdultWeekend'])
y = df.AdultWeekend
scaler2 = preprocessing.StandardScaler()
scaler2.fit(df2)
df2_scaled = scaler2.transform(df2)

In [103]:
y = y.ravel()
df2_train, df2_test, y_train, y_test = train_test_split(df2_scaled, y, test_size=0.25, random_state=1)

In [104]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(df2_train,y_train)

In [105]:
y_pred = model2.predict(df2_test)

In [106]:
print(explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred))

0.9243161723825675 5.502990808775245


In [107]:
model2.intercept_

64.03873572419111

In [108]:
pd.DataFrame(abs(lm2.coef_), df2.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
AdultWeekday         20.235205
summit_elev          12.612442
base_elev             9.791939
vertical_drop         4.367079
averageSnowfall       1.929559
quad                  1.554034
triple                1.385544
surface               1.255699
daysOpenLastYear      1.170290
Runs                  1.056548
fastQuads             0.726818
total_chairs          0.689683
projectedDaysOpen     0.468032
NightSkiing_ac        0.397652
fastSixes             0.373324
double                0.308310
SkiableTerrain_ac     0.294743
yearsOpen             0.283607
LongestRun_mi         0.220038
TerrainParks          0.190545
trams                 0.167374
fastEight             0.073238
Snow Making_ac        0.031914

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [109]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'base_elev', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'Alaska', 'Arizona', 'California', 'Colorado',
       'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Maine',
       'Maryland', 'Massachusetts', 'Michigan', 'Minnesota', 'Missouri',
       'Montana', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Dakota', 'Tennessee', 'Utah', 'Vermont',
       'Virginia', 'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype='object')

In [112]:
X3 = df.loc[: , 'trams':'NightSkiing_ac'].drop(columns=['AdultWeekend'])

In [113]:
X3.head()

trams  fastEight  fastSixes  fastQuads  quad  triple  double  surface  \
0      1        0.0          0          2     2       0       0        2   
1      0        0.0          0          0     0       0       4        0   
2      0        0.0          0          0     0       1       0        2   
3      0        0.0          1          0     2       2       1        2   
4      0        0.0          0          1     2       3       1        0   

   total_chairs  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
0             7  76.0           2.0            1.0             1610.0   
1             4  36.0           1.0            2.0              640.0   
2             3  13.0           1.0            1.0               30.0   
3             8  55.0           4.0            2.0              777.0   
4             7  65.0           2.0            1.2              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   projectedDaysOpen  NightSkiing_ac  
0              150.0           550.0  
1               90.0             0.0  
2              152.0            30.0  
3              122.0             0.0  
4              104.0            80.0

In [114]:
y = df.AdultWeekend
scaler3 = preprocessing.StandardScaler()
scaler3.fit(X3)
X3_scaled = scaler3.transform(X3)

In [115]:
y = y.ravel()

In [116]:
X3_train,X3_test,y_train,y_test = train_test_split(X3_scaled, y, test_size=0.25, random_state = 1)

In [117]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train, y_train)

In [118]:
y_pred = model3.predict(X3_test)

In [119]:
explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred)

(0.9262607076738476, 5.44363862673528)

In [120]:
model3.intercept_

64.03492371218334

In [121]:
pd.DataFrame(abs(model3.coef_), X3.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday         20.281218
averageSnowfall       2.066679
quad                  1.495971
surface               1.318477
triple                1.314431
daysOpenLastYear      1.099032
fastQuads             0.913233
Runs                  0.800082
total_chairs          0.700080
NightSkiing_ac        0.491846
double                0.296475
yearsOpen             0.269615
SkiableTerrain_ac     0.266588
fastSixes             0.239554
projectedDaysOpen     0.174575
Snow Making_ac        0.102174
LongestRun_mi         0.089217
TerrainParks          0.083403
fastEight             0.076708
trams                 0.039629

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.93486 | 5.07026 |-|
| Model 2. | 0.92432 | 5.50299 |'state'|
| Model 3. | 0.92626 | 5.44364 |'state','summit_elev','base_elev'|

Model Selection:

# Model 1 has the lowest Mean Abs Error.
# It also has the highest Explained Variance, closest to 1.0 than the others.
# Model 1 is the best